File: Angular/paragon/src/app/auth/auth-interceptor.ts
import { Injectable } from '@angular/core';
import { HttpRequest, HttpHandler, HttpEvent, HttpInterceptor } from '@angular/common/http';
import { Observable, from } from 'rxjs';
import { AuthService } from '@auth0/auth0-angular';
import { switchMap } from 'rxjs/operators';

@Injectable()
export class AuthInterceptor implements HttpInterceptor {

  constructor(private auth: AuthService) { }

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    // Only add an Authorization header if a token is available
    if(req.url.includes("assets/jobs.json")) return next.handle(req);
    if(req.url.includes(":3000")) return next.handle(req);
    return from(this.auth.getAccessTokenSilently()).pipe(
      switchMap(token => {
        // Clone the request and add the Authorization header if the token is available
        if (token) {
          const cloned = req.clone({
            setHeaders: {
              Authorization: `Bearer ${token}`
            }
          });
          return next.handle(cloned);
        }
        // If no token is available, pass the original request through
        return next.handle(req);
      })
    );
  }
}

File: Python/authInterceptor.py
import requests
from requests.adapters import HTTPAdapter

class AuthInterceptor(HTTPAdapter):
    def __init__(self, auth):
        self.auth = auth
        super().__init__()

    def send(self, request, **kwargs):
        # Inject the Authorization header if access_token is available
        access_token = self.auth.access_token
        if access_token:
            request.headers['Authorization'] = f'Bearer {access_token}'
            #print(f"{request.headers['Authorization']}")

        return super().send(request, **kwargs)
